HBSD: Make security.bsd.unprivileged_proc_debug per-jail

lattera opened the pull request:
Teach HardenedBSD to set security.bsd.unprivileged_proc_debug per-jail.
This is needed to create jails in which the Address Sanitizer (ASAN)
fully works as ASAN utilizes libkvm to inspect the virtual address
space. Instead of having to allow unprivileged process debugging for the
entire system, allow setting it on a per-jail basis.

The sysctl node is still security.bsd.unprivileged_proc_debug, but the
jail(8) param is hardening.unprivileged_proc_debug.

There's room for a bit more improvement here. One of HardenedBSD's
internal KPI functions, pax_feature_simple_validate_state, could use a
second argument to specify the default value in case of error. That work
should be done independent of this patch.

Signed-off-by:	Shawn Webb <shawn.webb@hardenedbsd.org>
Sponsored-by:	G2, Inc
github-issue:	#360

opntr-auto commented:
Hi!

In overall the idea isn't bad, but not really fits in the current
HardenedBSD control model. It's possible to force them in, but it's quirky.

AFAIK it would be better to implement them under the FreeBSD's
"kern_prot.c" namespace and try to upstream them. If they didn't accept
them, then we cloud stay on this or like this implementation.

On Thursday, November 22, 2018, Shawn Webb <notifications@github.com> wrote:

> Teach HardenedBSD to set security.bsd.unprivileged_proc_debug per-jail.
> This is needed to create jails in which the Address Sanitizer (ASAN)
> fully works as ASAN utilizes libkvm to inspect the virtual address
> space. Instead of having to allow unprivileged process debugging for the
> entire system, allow setting it on a per-jail basis.
>
> The sysctl node is still security.bsd.unprivileged_proc_debug, but the
> jail(8) param is hardening.unprivileged_proc_debug.
>
> There's room for a bit more improvement here. One of HardenedBSD's
> internal KPI functions, pax_feature_simple_validate_state, could use a
> second argument to specify the default value in case of error. That work
> should be done independent of this patch.
>
> Signed-off-by: Shawn Webb shawn.webb@hardenedbsd.org
> Sponsored-by: G2, Inc
> github-issue: #360 <https://github.com/HardenedBSD/hardenedBSD/issues/360>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/HardenedBSD/hardenedBSD/pull/361
> Commit Summary
>
>    - HBSD: Make security.bsd.unprivileged_proc_debug per-jail
>
> File Changes
>
>    - *M* sys/hardenedbsd/hbsd_pax_hardening.c
>    <https://github.com/HardenedBSD/hardenedBSD/pull/361/files#diff-0>
>    (63)
>    - *M* sys/kern/kern_prot.c
>    <https://github.com/HardenedBSD/hardenedBSD/pull/361/files#diff-1>
>    (22)
>    - *M* sys/sys/pax.h
>    <https://github.com/HardenedBSD/hardenedBSD/pull/361/files#diff-2> (1)
>
> Patch Links:
>
>    - https://github.com/HardenedBSD/hardenedBSD/pull/361.patch
>    - https://github.com/HardenedBSD/hardenedBSD/pull/361.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/HardenedBSD/hardenedBSD/pull/361>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AILQ0eDUEoN0OhuIFc6NdW062Fl9tJI-ks5uxwjAgaJpZM4Yv3Wj>
> .
>

opntr-auto commented:
In the same time as the patch is in now, we loose a relevant and important
comment from the code.

On Thursday, November 22, 2018, Oliver Pinter <oliver.pinter@hardenedbsd.org>
wrote:

> Hi!
>
> In overall the idea isn't bad, but not really fits in the current
> HardenedBSD control model. It's possible to force them in, but it's quirky.
>
> AFAIK it would be better to implement them under the FreeBSD's
> "kern_prot.c" namespace and try to upstream them. If they didn't accept
> them, then we cloud stay on this or like this implementation.
>
> On Thursday, November 22, 2018, Shawn Webb <notifications@github.com>
> wrote:
>
>> Teach HardenedBSD to set security.bsd.unprivileged_proc_debug per-jail.
>> This is needed to create jails in which the Address Sanitizer (ASAN)
>> fully works as ASAN utilizes libkvm to inspect the virtual address
>> space. Instead of having to allow unprivileged process debugging for the
>> entire system, allow setting it on a per-jail basis.
>>
>> The sysctl node is still security.bsd.unprivileged_proc_debug, but the
>> jail(8) param is hardening.unprivileged_proc_debug.
>>
>> There's room for a bit more improvement here. One of HardenedBSD's
>> internal KPI functions, pax_feature_simple_validate_state, could use a
>> second argument to specify the default value in case of error. That work
>> should be done independent of this patch.
>>
>> Signed-off-by: Shawn Webb shawn.webb@hardenedbsd.org
>> Sponsored-by: G2, Inc
>> github-issue: #360
>> <https://github.com/HardenedBSD/hardenedBSD/issues/360>
>> ------------------------------
>> You can view, comment on, or merge this pull request online at:
>>
>>   https://github.com/HardenedBSD/hardenedBSD/pull/361
>> Commit Summary
>>
>>    - HBSD: Make security.bsd.unprivileged_proc_debug per-jail
>>
>> File Changes
>>
>>    - *M* sys/hardenedbsd/hbsd_pax_hardening.c
>>    <https://github.com/HardenedBSD/hardenedBSD/pull/361/files#diff-0>
>>    (63)
>>    - *M* sys/kern/kern_prot.c
>>    <https://github.com/HardenedBSD/hardenedBSD/pull/361/files#diff-1>
>>    (22)
>>    - *M* sys/sys/pax.h
>>    <https://github.com/HardenedBSD/hardenedBSD/pull/361/files#diff-2>
>>    (1)
>>
>> Patch Links:
>>
>>    - https://github.com/HardenedBSD/hardenedBSD/pull/361.patch
>>    - https://github.com/HardenedBSD/hardenedBSD/pull/361.diff
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/HardenedBSD/hardenedBSD/pull/361>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AILQ0eDUEoN0OhuIFc6NdW062Fl9tJI-ks5uxwjAgaJpZM4Yv3Wj>
>> .
>>
>

lattera committed:
HBSD: Make security.bsd.unprivileged_proc_debug per-jail

Teach HardenedBSD to set security.bsd.unprivileged_proc_debug per-jail.
This is needed to create jails in which the Address Sanitizer (ASAN)
fully works as ASAN utilizes libkvm to inspect the virtual address
space. Instead of having to allow unprivileged process debugging for the
entire system, allow setting it on a per-jail basis.

The sysctl node is still security.bsd.unprivileged_proc_debug, but the
jail(8) param is hardening.unprivileged_proc_debug.

There's room for a bit more improvement here. One of HardenedBSD's
internal KPI functions, pax_feature_simple_validate_state, could use a
second argument to specify the default value in case of error. That work
should be done independent of this patch.

Signed-off-by:	Shawn Webb <shawn.webb@hardenedbsd.org>
Sponsored-by:	G2, Inc
github-issue:	#360

lattera commented:
Sure. I can figure out a different, upstreamable implementation. Perhaps using the new dynamic allow.* framework.

lattera commented:
The other patch has been accepted by FreeBSD and is awaiting committal.
